tests/components/smappee/test_config_flow.py

```diff
@@ -1,16 +1,326 @@
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {"environment": ENV_LOCAL}
+    )
+
+    assert result["step_id"] == ENV_LOCAL
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+
+async def test_show_zeroconf_connection_error_form(hass):
+    """Test that the zeroconf confirmation form is served."""
+    with patch("pysmappee.api.SmappeeLocalApi.logon", return_value=None):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_ZEROCONF},
+            data={
+                "host": "1.2.3.4",
+                "port": 22,
+                CONF_HOSTNAME: "Smappee1006000212.local.",
+                "type": "_ssh._tcp.local.",
+                "name": "Smappee1006000212._ssh._tcp.local.",
+                "properties": {"_raw": {}},
+            },
+        )
+
+        assert result["description_placeholders"] == {CONF_SERIALNUMBER: "1006000212"}
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "zeroconf_confirm"
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": "1.2.3.4"}
+        )
+
+        assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+        assert result["reason"] == "connection_error"
+        assert len(hass.config_entries.async_entries(DOMAIN)) == 0
+
+
+async def test_connection_error(hass):
+    """Test we show user form on Smappee connection error."""
+    with patch("pysmappee.api.SmappeeLocalApi.logon", return_value=None):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER},
+        )
+        assert result["step_id"] == "environment"
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"environment": ENV_LOCAL}
+        )
+        assert result["step_id"] == ENV_LOCAL
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+        result = await hass.config_entries.flow.async_configure(
+            result["flow_id"], {"host": "1.2.3.4"}
+        )
+        assert result["reason"] == "connection_error"
+        assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+
+
+async def test_zeroconf_wrong_mdns(hass):
+    """Test we abort if unsupported mDNS name is discovered."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data={
+            "host": "1.2.3.4",
+            "port": 22,
+            CONF_HOSTNAME: "example.local.",
+            "type": "_ssh._tcp.local.",
+            "name": "example._ssh._tcp.local.",
+            "properties": {"_raw": {}},
+        },
+    )
+
+    assert result["reason"] == "invalid_mdns"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)